hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/recovery/ReconOmMetadataManagerImpl.java

```diff
@@ -147,4 +157,158 @@ public long getLastSequenceNumberFromDB() {
+   * This method can be optimized by using username as a filter.
+   * @return a list of volume names under the system
+   */
+  @Override
+  public List<OmVolumeArgs> listVolumes() throws IOException {
+    return listVolumes(null, Integer.MAX_VALUE);
+  }
+
+  @Override
+  public boolean volumeExists(String volName) throws IOException {
+    String volDBKey = getVolumeKey(volName);
+    return getVolumeTable().getSkipCache(volDBKey) != null;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public List<OmBucketInfo> listBucketsUnderVolume(final String volumeName,
+       final String startBucket, final int maxNumOfBuckets) throws IOException {
+    List<OmBucketInfo> result = new ArrayList<>();
+
+    // List all buckets if volume is empty
+    if (Strings.isNullOrEmpty(volumeName)) {
+      // startBucket requires the knowledge of the volume, for example
+      // there might be buckets with the same names under different
+      // volumes. Hence, startBucket will be ignored if
+      // volumeName is not specified
+      return listAllBuckets(maxNumOfBuckets);
+    }
+
+    if (!volumeExists(volumeName)) {
+      return result;
+    }
+
+    String startKey;
+    boolean skipStartKey = false;
+    if (StringUtil.isNotBlank(startBucket)) {
+      startKey = getBucketKey(volumeName, startBucket);
+      skipStartKey = true;
+    } else {
+      startKey = getBucketKey(volumeName, null);
+    }
+
+    String seekPrefix = getVolumeKey(volumeName + OM_KEY_PREFIX);
+
+    int currentCount = 0;
+
+    // Unlike in {@link OmMetadataManagerImpl}, the buckets are queried directly
+    // from the volume table (not through cache) since Recon does not use
+    // Table cache.
+    try (TableIterator<String, ? extends Table.KeyValue<String, OmBucketInfo>>
+        iterator = getBucketTable().iterator(seekPrefix)) {
+
+      while (currentCount < maxNumOfBuckets && iterator.hasNext()) {
+        Table.KeyValue<String, OmBucketInfo> kv =
+            iterator.next();
+
+        String key = kv.getKey();
+        OmBucketInfo omBucketInfo = kv.getValue();
+
+        if (omBucketInfo != null) {
+          if (key.equals(startKey) && skipStartKey) {
+            continue;
+          }
+
+          // We should return only the keys, whose keys match with prefix and
+          // the keys after the startBucket.
+          if (key.startsWith(seekPrefix) && key.compareTo(startKey) >= 0) {
+            result.add(omBucketInfo);
+            currentCount++;
+          }
+        }
+      }
+    }
+    return result;
+  }
+
+  /**
+   * List all buckets under a volume, if volume name is null, return all buckets
+   * under the system.
+   * @param volumeName volume name without protocol prefix
+   * @return buckets under volume or all buckets if volume is null
+   * @throws IOException IOE
+   */
+  public List<OmBucketInfo> listBucketsUnderVolume(final String volumeName)
+      throws IOException {
+    return listBucketsUnderVolume(volumeName, null,
```

Comment:
Still I feel max 1000 should be default and UI should show a message.

Reply:
Sorry for omitting this detail, but I added editable drop downs in the UI for user to be able to change the maximum number of volumes/buckets fetched from the backend.

<img width="618" alt="image" src="https://github.com/apache/ozone/assets/36403683/9466208a-8319-47dd-b410-e41333c383ef">

For example in `volumes.tsx`

```typescript
<CreatableSelect
    className='multi-select-container'
    isClearable={false}
    isDisabled={loading}
    isLoading={loading}
    onChange={this._handleLimitChange}
    onCreateOption={this._onCreateOption}
    isValidNewOption={(input, value, _option) => {
      // Only number will be accepted
      return !isNaN(parseInt(input))
    }}
    options={LIMIT_OPTIONS}
    hideSelectedOptions={false}
    value={selectedLimit}
    createOptionPosition='last'
    formatCreateLabel={(input) => {
      return `new limit... ${input}`
    }}
/> Limit
```

Reply:
If user needs to load more volumes / buckets, they can change the limit / create a new value on top of the default ones.

Reply:
> Hence, currently, my current suggestion is that we can have a drop down that will set the limit parameter and initialize it to a good default (e.g. 5000) to bound the number of buckets/volumes fetched

As I mentioned in the earlier thread, currently it's very hard to implement pagination in RocksDB without sacrificing sortability of different fields. So to bound the number of objects returned, we set the `limit` when fetching from the Recon API.

The pagination implementation can be implemented in different ticket, but I think the most straightaway solution is to write a task to get all he  volumes / buckets from the RocksDB, creating separate derby tables for volumes and buckets, and probably wrapping it in JPA (supported natively) / JOOQ (might need to write sql for pagination). In that way, we can achieve the expected pagination capabilities easily using known SQL methods (e.g. offset and limit)

Reply:
@ivandika3 Thanks for considering the comments. Patch looks good in current state, however as per your last comment regarding use of derby DB, I am not pretty sure about the performance when large number of buckets, but yeah while implementing, you can have a small perf test. Rest all looks good. +1

Reply:
@devmadhuu Thank you for the detailed reviews.